cmd/gen-go-sample/inittree.go

```diff
@@ -76,28 +81,68 @@ func (t *initTree) get(k string, info pbinfo.Info) (*initTree, error) {
 			continue
 		}
 
-		if tn := f.GetTypeName(); tn != "" {
-			typ := info.Type[f.GetTypeName()]
-			if typ == nil {
-				return nil, errors.E(nil, "cannot find descriptor of %q", f.GetTypeName())
-			}
-			v.typ.desc = typ
-		} else {
+		if tn := f.GetTypeName(); tn == "" {
+			// We're a primitive type.
+
 			// Since the tokens are given to us by scanner, they must already be a valid token of some type,
 			// no need to check exhaustively.
 			switch f.GetType() {
 			case descriptor.FieldDescriptorProto_TYPE_BOOL:
-				v.typ.primValid = func(s string) bool { return s == "true" || s == "false" }
+				v.typ.valValid = func(s string) bool { return s == "true" || s == "false" }
 			case descriptor.FieldDescriptorProto_TYPE_BYTES, descriptor.FieldDescriptorProto_TYPE_STRING:
-				v.typ.primValid = func(s string) bool { return s[0] == '"' }
+				v.typ.valValid = func(s string) bool { return s[0] == '"' }
 			case descriptor.FieldDescriptorProto_TYPE_DOUBLE, descriptor.FieldDescriptorProto_TYPE_FLOAT:
-				v.typ.primValid = func(s string) bool { return s == "inf" || s == "nan" || strings.Trim(s, "+-.0123456789") == "" }
+				v.typ.valValid = func(s string) bool { return s == "inf" || s == "nan" || strings.Trim(s, "+-.0123456789") == "" }
 			default:
-				v.typ.primValid = func(s string) bool { return strings.Trim(s, "0123456789") == "" }
+				v.typ.valValid = func(s string) bool { return strings.Trim(s, "0123456789") == "" }
 			}
+		} else if typ := info.Type[tn]; typ == nil {
+			return nil, errors.E(nil, "cannot find descriptor of %q", f.GetTypeName())
+		} else if enum, ok := typ.(*descriptor.EnumDescriptorProto); ok {
```

Comment:
Would it make sense to split the part that checks if the value is valid into a separate function? This function seems quite large already.